### PR TITLE
Correctness for x86_64 routines

### PR DESCRIPTION
**Pull request by chameco**
Currently:
- Routines in `add_mod_256-x86_64.pl`, `add_mod_384-x86_64.pl`, and `add_mod_384x384-x86_64.pl` are proved equivalent to fairly high-level Cryptol functions defined over bitvectors. For example, `add_mod_384` is proved equivalent to:
```
vec384_rep ((vec384_abs a + vec384_abs b) % vec384_abs modulus384)
```
- Routines in `mulx_mont_256-x86_64.pl`, `mulx_mont_384-x86_64.pl`, and `ctx_inverse_mod_384-x86_64.pl` are proved equivalent to Cryptol functions that closely match the structure of the x86_64 code. For example, `sqrx_mont_384x` is proved equivalent to:
```
sqrx_mont_384x : Vec768 -> Vec384 -> [64] -> Vec768
sqrx_mont_384x a m n0 = result
  where
    t0 = __add_mod_384 (take a) (drop a) m
    t1 = __sub_mod_384 (take a) (drop a) m
    tmp = mulx_mont_384 (take a) (drop a) m n0
    ret_im = __add_mod_384 tmp tmp m
    ret_re = mulx_mont_384 t0 t1 m n0
    result = ret_re # ret_im
```
(where `__add_mod_384` etc. are verbose Cryptol functions that closely match the x86_64 helper routines with the same names).
- No proofs for `mulq` variants
- `mul_by_b_onE2` proof has poor solver performance and is admitted

### CODE REVIEW

**Comment by chameco** [comment]
pinging @andreistefanescu 

The CI failure should be fixed once the SAW build is updated to include GaloisInc/saw-script@aac0a02121b05d3e93331d4cfb21ecefc461ad6a.

File: proof/x86/correctness.saw [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+let verify_x86_noadx = verify_x86_helper "../../../build/x86_noadx/libblst.so" m_noadx;
+let verify_x86_recent = verify_x86_helper "../../../build/x86_recent/libblst.so" m_recent;
+let verify_x86_noadx_recent = verify_x86_helper "../../../build/x86_noadx_recent/libblst.so" m_noadx_recent;
+
+enable_what4_hash_consing;
+enable_x86_what4_hash_consing;
+
+include "correctness/helpers.saw";
+
+//include "correctness/add_mod_256.saw";
```
Comment (andreistefanescu) [inline proof/x86/correctness.saw:29]: nit: remove commented lines

File: proof/x86/correctness/add_mod_256.saw [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  crucible_points_to ret_ptr (crucible_term new_ret);
+};
+
+///////////////////////////////////////////////////////////////////////////////
+// Proofs
+///////////////////////////////////////////////////////////////////////////////
+
+// add_mod_256
+verify_x86 "add_mod_256" add_mod_256_spec w4;
+// verify_x86 "add_mod_256" add_mod_256_alias_ret_a_spec w4;
```
Comment (andreistefanescu) [inline proof/x86/correctness/add_mod_256.saw:187]: nit: let's remove the commented lines, alternatively, wrap with `do_prove = false; ...; do_prove = true;`

File: proof/x86/correctness/add_mod_384x384.saw [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  crucible_points_to ret_ptr (crucible_term new_ret);
+};
+
+///////////////////////////////////////////////////////////////////////////////
+// Proofs
+///////////////////////////////////////////////////////////////////////////////
+
+// add_mod_384x384
+verify_x86 "add_mod_384x384" add_mod_384x384_spec w4;
+// verify_x86 "add_mod_384x384" add_mod_384x384_alias_ret_a_spec w4;
```
Comment (andreistefanescu) [inline proof/x86/correctness/add_mod_384x384.saw:122]: nit: same about commented lines

File: proof/x86/correctness/ctx_inverse_mod_384.saw [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+//   directly to the offset. From an assurance perspective, this corresponds to
+//   a loss of some generality in the precondition describing the initial value
+//   of RSP.
+
+set_x86_stack_base_align 512;
+
+// ctx_inverse_mod_383
+verify_x86_recent "ctx_inverse_mod_383" ct_inverse_mod_383_spec do {
+  // quickcheck 10;
+  goal_eval_unint [];
```
Comment (andreistefanescu) [inline proof/x86/correctness/ctx_inverse_mod_384.saw:45]: neat, I like it

File: proof/x86/correctness/add_mod_384.saw [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+  crucible_points_to ret_ptr (crucible_term ret);
+};
+
+///////////////////////////////////////////////////////////////////////////////
+// Proofs
+///////////////////////////////////////////////////////////////////////////////
+
+// add_mod_384
+verify_x86 "add_mod_384" add_mod_384_spec w4;
+// verify_x86 "add_mod_384" add_mod_384_alias_ret_a_spec w4;
```
Comment (andreistefanescu) [inline proof/x86/correctness/add_mod_384.saw:518]: nit: same about commented lines

File: proof/x86/correctness/add_mod_384.saw [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+// mul_by_b_onE2
+let mul_by_b_onE2_spec = do {
+  crucible_alloc_global "BLS12_381_P";
+  crucible_points_to
+    (crucible_global "BLS12_381_P")
+    (crucible_term modulus384);
+
+  out_ptr <- crucible_alloc vec384x_type;
+
+  // in_ptr <- crucible_alloc vec384x_type;
```
Comment (andreistefanescu) [inline proof/x86/correctness/add_mod_384.saw:262]: nit: remove

**Review by andreistefanescu — APPROVED** [review]
@chameco I don't know what's an appropriate comment for this work... neat, well done, rocket emoji? Thank you for doing it!

**Pull request merged**